agent/agent.go

```diff
@@ -218,6 +213,7 @@ func (a *Agent) run(ctx context.Context) {
 			if a.err == nil {
 				a.err = ctx.Err()
 			}
+			session.close()
```

Comment:
> I always get confused about where it's safe to call these close functions and where it's not safe.

This modeled after a critical section. For the most part, methods that are un-exported are safe to call.

In this case, `session.close()` is protected by the critical section.

I do want to find a better way to organize this code but I am not sure the event model is quite right. For the most part, there are handlers for each section, but other parts modify local loop state, so they need to close over those state variables. The `session` is primary resource managed in this for loop, so that is why we see this here.